### PR TITLE
[stdlib] Apply @_transparent consistently for (U)Int128

### LINKED ISSUE #74481: [SE-0425] `Int128` and `UInt128` APIs have different attributes
For the following APIs:

* [`Int128`](https://github.com/apple/swift/blob/main/stdlib/public/core/Int128.swift) uses either `@_transparent`, or `@inline(__always)`, or no attribute.

* [`UInt128`](https://github.com/apple/swift/blob/main/stdlib/public/core/UInt128.swift) uses `@inlinable` only.

| APIs                                      | `Int128`            | `UInt128`    |
| :---------------------------------------- | :------------------ | :----------- |
| `zero`, `min`, `max`                      | `@_transparent`     | `@inlinable` |
| `init(_builtinIntegerLiteral:)`           | `@_transparent`     | `@inlinable` |
| `addingReportingOverflow(_:)`             | `@_transparent`     | `@inlinable` |
| `subtractingReportingOverflow(_:)`        | `@_transparent`     | `@inlinable` |
| `multipliedReportingOverflow(by:)`        | `@inline(__always)` | `@inlinable` |
| `dividedReportingOverflow(by:)`           | `@inline(__always)` | `@inlinable` |
| `remainderReportingOverflow(dividingBy:)` | `@inline(__always)` | `@inlinable` |
| `*`, `/`, `%`                             |                     | `@inlinable` |
| `&=`, `\|=`, `^=`, `&>>=`, `&<<=`         |                     | `@inlinable` |
| `trailingZeroBitCount`                    |                     | `@inlinable` |
| `bitWidth`                                | `@_transparent`     | `@inlinable` |
| `nonzeroBitCount`                         |                     | `@inlinable` |
| `leadingZeroBitCount`                     |                     | `@inlinable` |
| `byteSwapped`                             |                     | `@inlinable` |

Should the attributes be the same, when the APIs have the same implementations?

(`main` and `release/6.0` branches.)

### PR DESCRIPTION
**Pull request by Azoy**
Also adds a concrete implementation of `_lowWord` for both `Int128` and `UInt128`. I've sprinkled available attributes everywhere because I personally dislike the implicit availability from the extension declaration because a new API introduced in an existing extension may forget to add availability which would be incorrect, so littering this everywhere hopefully makes it more obvious.

Resolves: https://github.com/apple/swift/issues/74481

### CODE REVIEW

**Comment by Azoy** [comment]
@swift-ci please test

**Comment by Azoy** [comment]
@swift-ci please benchmark

File: stdlib/public/core/Int128.swift [hunk 1]
```diff
@@ -250,21 +288,23 @@ extension Int128 {
...
-  
-  @inline(__always)
+
+  @available(SwiftStdlib 6.0, *)
+  @_transparent
   public func remainderReportingOverflow(
     dividingBy other: Self
   ) -> (partialValue: Self, overflow: Bool) {
-    precondition(other != .zero, "Remainder dividing by zero.")
+    _precondition(other != .zero, "Division by zero in remainer operation")
```
Comment (stephentyrone) [inline stdlib/public/core/Int128.swift:307]:
```suggestion
    _precondition(other != .zero, "Division by zero in remainder operation")
```

File: stdlib/public/core/Int128.swift [hunk 2]
```diff
@@ -334,59 +402,90 @@ extension Int128: SignedNumeric {
...
+  @_transparent
   public var trailingZeroBitCount: Int {
     _low == 0 ? 64 + _high.trailingZeroBitCount : _low.trailingZeroBitCount
   }
+
+  @available(SwiftStdlib 6.0, *)
+  @_transparent
+  public var _lowWord: UInt {
+#if _pointerBitWidth(_64)
+    UInt(_low._value)
```
Comment (stephentyrone) [inline stdlib/public/core/Int128.swift:453]: Why isn't this just `_low`?

File: stdlib/public/core/Int128.swift [hunk 3]
```diff
@@ -334,59 +402,90 @@ extension Int128: SignedNumeric {
...
+#elseif _pointerBitWidth(_32)
+    UInt(Builtin.trunc_Int64_Int32(_low._value))
+#else
+    UInt(truncatingIfNeeded: _low)
```
Comment (stephentyrone) [inline stdlib/public/core/Int128.swift:457]: Formally I think the `#else` branch here should probably hit the default implementation instead, but that requires refactoring this a bit, so I'm ok with this for now.

File: stdlib/public/core/UInt128.swift [hunk 4]
```diff
@@ -229,21 +268,23 @@ extension UInt128 {
...
-  
-  @inlinable
+
+  @available(SwiftStdlib 6.0, *)
+  @_transparent
   public func remainderReportingOverflow(
     dividingBy other: Self
   ) -> (partialValue: Self, overflow: Bool) {
-    precondition(other != .zero, "Remainder dividing by zero.")
+    _precondition(other != .zero, "Division by zero in remainer operation")
```
Comment (stephentyrone) [inline stdlib/public/core/UInt128.swift:287]:
```suggestion
    _precondition(other != .zero, "Division by zero in remainder operation")
```

**Review by stephentyrone — CHANGES_REQUESTED** [review]

File: stdlib/public/core/Int128.swift [hunk 5]
```diff
@@ -250,21 +288,23 @@ extension Int128 {
       return (partialValue, overflow || partialValue < 0)
     }
   }
-  
-  @inline(__always)
+
+  @available(SwiftStdlib 6.0, *)
+  @_transparent
```
Comment (stephentyrone) [inline stdlib/public/core/Int128.swift:293]: I'm not certain that we want these to be `@_transparent`. These are potentially pretty heavyweight operations to drop into the caller (though we might reap benefits in the case of constant divisors).

**Review by stephentyrone — APPROVED** [review]

**Comment by Azoy** [comment]
@stephentyrone how do we feel about making `_low` and `_high` public? Considering that the underlying builtin `_value` property is public (for good reason) it seems like we should just make these other accessors public as well. Thoughts?

**Comment by stephentyrone** [comment]
It's a little bit weird since we don't have them for any other integer types, but other than that I'm OK with it.

**Comment by Azoy** [comment]
@swift-ci please test

**Comment by Azoy** [comment]
@swift-ci please benchmark

File: stdlib/public/core/Int128.swift (hunk 3 above)
Comment (Azoy) [reply stdlib/public/core/Int128.swift]: Yeah this implementation is technically incorrect for words larger than 64 bits, so I just wrote this as truncating self.

File: stdlib/public/core/Int128.swift [hunk 6]
```diff
@@ -334,59 +398,90 @@ extension Int128: SignedNumeric {
...
+  public var _lowWord: UInt {
+#if _pointerBitWidth(_64)
+    UInt(Builtin.trunc_Int128_Int64(_value))
+#elseif _pointerBitWidth(_32)
+    UInt(Builtin.trunc_Int128_Int32(_value))
+#else
+    UInt(truncatingIfNeeded: self)
```
[the remark this answers is not in the document]
Comment (Azoy) [reply stdlib/public/core/Int128.swift]: Yep, I think we should just figure out what to do when those platforms come around, so for now I've just emitted a static error.

**Comment by Azoy** [comment]
@swift-ci please test

**Comment by Azoy** [comment]
@swift-ci please test

**Comment by Azoy** [comment]
@swift-ci test WebAssembly

**Comment by Azoy** [comment]
Sure!

**Pull request merged**